ably/realtime_presence.go

```diff
@@ -25,11 +26,36 @@ type RealtimePresence struct {
 }
 
 func newRealtimePresence(channel *RealtimeChannel) *RealtimePresence {
-	return &RealtimePresence{
+	pres := &RealtimePresence{
 		subs:    newSubscriptions(subscriptionPresenceMessages),
 		channel: channel,
+		event:   make(chan State, 1),
 		members: make(map[string]*proto.PresenceMessage),
 	}
+	pres.syncMtx.Lock() // initially we lock waiting for channel's attached state
+	channel.On(pres.event)
+	go pres.processIncomingEvents()
+	return pres
+}
+
+func (pres *RealtimePresence) processIncomingEvents() {
+	for event := range pres.event {
+		switch event.State {
+		case StateChanAttached:
+			pres.mtx.Lock()
+			if !pres.syncInProgress {
+				// channel is attached and there's no sync in progress
+				pres.syncMtx.Unlock()
+			}
+			pres.mtx.Unlock()
+		case StateChanDetached, StateChanFailed, StateChanClosed:
+			pres.syncMtx.Lock()
```

Comment:
You are locking this in a completely separated thread but I don't see where it would get unlocked.